workflow/util/util.go

```diff
@@ -853,178 +1218,152 @@ func FormulateRetryWorkflow(ctx context.Context, wf *wfv1.Workflow, restartSucce
-	newWF.Status.Nodes = make(wfv1.Nodes)
-	newWF.Status.Message = ""
-	newWF.Status.StartedAt = metav1.Time{Time: time.Now().UTC()}
-	newWF.Status.FinishedAt = metav1.Time{}
-	if newWF.Status.StoredWorkflowSpec != nil {
-		newWF.Status.StoredWorkflowSpec.Shutdown = ""
+	newWf, err := createNewRetryWorkflow(wf, parameters)
+	if err != nil {
+		return nil, nil, err
 	}
-	newWF.Spec.Shutdown = ""
-	newWF.Status.PersistentVolumeClaims = []apiv1.Volume{}
-	if newWF.Spec.ActiveDeadlineSeconds != nil && *newWF.Spec.ActiveDeadlineSeconds == 0 {
-		// if it was terminated, unset the deadline
-		newWF.Spec.ActiveDeadlineSeconds = nil
+
+	deleteNodesMap, err := getNodeIDsToResetNoChildren(restartSuccessful, nodeFieldSelector, wf.Status.Nodes)
+	if err != nil {
+		return nil, nil, err
 	}
-	// Override parameters
-	if parameters != nil {
-		if _, ok := wf.ObjectMeta.Labels[common.LabelKeyPreviousWorkflowName]; ok {
-			log.Warnln("Overriding parameters on resubmitted workflows may have unexpected results")
+
+	failed := make(map[string]bool)
+	for nodeID, node := range wf.Status.Nodes {
+		if node.Phase.FailedOrError() && shouldRetryFailedType(node.Type) && !isDescendantNodeSucceeded(wf, node, deleteNodesMap) {
+			failed[nodeID] = true
 		}
-		err := overrideParameters(newWF, parameters)
+	}
+	for failedNode := range failed {
+		deleteNodesMap[failedNode] = true
+	}
+
+	nodes, err := newWorkflowsDag(wf)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	toReset := make(map[string]bool)
+	toDelete := make(map[string]bool)
+
+	nodesMap := make(map[string]*node)
+	for i := range nodes {
+		nodesMap[nodes[i].n.ID] = nodes[i]
+	}
+
+	nodes = dagSortedNodes(nodes, wf.Name)
+
+	deleteNodes := make([]*node, 0)
+
+	// deleteNodes will not contain an exit node
+	for i := range nodes {
+		if _, ok := deleteNodesMap[nodes[i].n.ID]; ok {
+			deleteNodes = append(deleteNodes, nodes[i])
+		}
+	}
+
+	// this is kind of complex
+	// we rely on deleteNodes being topologically sorted.
+	// this is done via a breadth first search on the dag via `dagSortedNodes`
+	// This is because nodes at the top take precedence over nodes at the bottom.
+	// if a failed node was declared to be scheduled for deletion, we should
+	// never execute resetPath on that node.
+	// we ensure this behaviour via calling resetPath in topological order.
+	for i := range deleteNodes {
+		currNode := deleteNodes[i]
+		shouldDelete := toDelete[currNode.n.ID]
+		if shouldDelete {
+			continue
+		}
+		pathToReset, pathToDelete, err := resetPath(nodes, currNode.n.ID)
 		if err != nil {
 			return nil, nil, err
 		}
+		toReset = setUnion(toReset, pathToReset)
+		toDelete = setUnion(toDelete, pathToDelete)
 	}
 
-	onExitNodeName := wf.ObjectMeta.Name + ".onExit"
-	// Get all children of nodes that match filter
-	nodeIDsToReset, err := getNodeIDsToReset(restartSuccessful, nodeFieldSelector, wf.Status.Nodes)
-	if err != nil {
-		return nil, nil, err
+	for k := range toReset {
```

Comment:
Swap `k` for `nodeName`. Short variable names for array indices are OK, but here it's a map key and more confusing if not named for what it is.